core/domain/exp_services.py

```diff
@@ -2047,3 +2052,56 @@ def get_interaction_id_for_state(exp_id, state_name):
         return exploration.get_interaction_id_by_state_name(state_name)
     raise Exception(
         'There exist no state in the exploration with the given state name.')
+
+
+def rollback_exploration_to_safe_state(exp_id):
+    """Rolls back exploration to the latest state where related meta models are
+    valid.
+
+    Args:
+        exp_id: str. The ID of the exp.
+
+    Returns:
+        str. The version of the exploration.
+    """
+    exploration_model = exp_models.ExplorationModel.get(exp_id)
+    current_version_in_exp_model = exploration_model.version
+    last_known_safe_version = exploration_model.version
+    snapshot_content_model = None
+    snapshot_metadata_model = None
+    for version in range(current_version_in_exp_model, 1, -1):
+        snapshot_content_model = (
+            exp_models.ExplorationSnapshotContentModel.get(
+                '%s-%s' % (exp_id, version), strict=False))
+        snapshot_metadata_model = (
+            exp_models.ExplorationSnapshotMetadataModel.get(
+                '%s-%s' % (exp_id, version), strict=False))
+        if snapshot_content_model is None and snapshot_metadata_model is None:
+            last_known_safe_version = version - 1
+        elif (
+            snapshot_content_model is None and
+            snapshot_metadata_model is not None
+        ):
+            snapshot_metadata_model.delete()
```

Comment:
Technically you should probably collect the model keys and delete them all at the same time as you fix the exp model. Just in case some error happens in between to result in an inconsistent state.

I suppose this may not be necessary for the specific case we've identified, but if we're planning to program this as a general thing, that might be the better way to do it.